File: app/src/main/java/me/zipi/navitotesla/background/TokenWorker.java
package me.zipi.navitotesla.background;

import android.content.Context;
import android.util.Log;

import java.util.concurrent.TimeUnit;

import androidx.annotation.NonNull;
import androidx.work.Constraints;
import androidx.work.ExistingPeriodicWorkPolicy;
import androidx.work.NetworkType;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;
import androidx.work.Worker;
import androidx.work.WorkerParameters;
import me.zipi.navitotesla.model.Token;
import me.zipi.navitotesla.service.NaviToTeslaService;
import me.zipi.navitotesla.util.AnalysisUtil;
import me.zipi.navitotesla.util.PreferencesUtil;

public class TokenWorker extends Worker {

    private static String workName = "refreshTokenWorker";

    public TokenWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    public static void startBackgroundWork(@NonNull Context context) {
        if (PreferencesUtil.loadToken(context) == null) {
            AnalysisUtil.log("Token is empty. add work ignore");
            return;
        }
        AnalysisUtil.log("Add background refresh token");
        PeriodicWorkRequest workRequest = new PeriodicWorkRequest.Builder(TokenWorker.class, 15, TimeUnit.MINUTES)
                .setConstraints(new Constraints.Builder()
                        .setRequiredNetworkType(NetworkType.CONNECTED)
                        .build())
                .build();
        WorkManager.getInstance(context).enqueueUniquePeriodicWork(
                workName,
                ExistingPeriodicWorkPolicy.KEEP,
                workRequest);
    }

    public static void cancelBackgroundWork(@NonNull Context context) {
        try {
            WorkManager.getInstance(context).cancelUniqueWork(workName);
        } catch (Exception e) {
            AnalysisUtil.recordException(e);
        }
    }

    @NonNull
    @Override
    public Result doWork() {

        Log.i(TokenWorker.class.getName(), "Start background refresh token");
        AnalysisUtil.log("Start background refresh token");
        NaviToTeslaService service = new NaviToTeslaService(getApplicationContext());
        Token token = service.refreshToken();

        if (token != null) {
            return Result.success();
        } else {
            return Result.failure();
        }
    }
}

File: app/src/main/kotlin/me/zipi/navitotesla/background/TokenWorker.kt
package me.zipi.navitotesla.background

import android.content.Context
import android.util.Log
import androidx.work.Constraints
import androidx.work.CoroutineWorker
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.NetworkType
import androidx.work.PeriodicWorkRequest
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.WorkerParameters
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import me.zipi.navitotesla.service.NaviToTeslaService
import me.zipi.navitotesla.util.AnalysisUtil
import me.zipi.navitotesla.util.PreferencesUtil
import java.util.concurrent.TimeUnit

class TokenWorker(
    context: Context,
    workerParams: WorkerParameters,
) : CoroutineWorker(context, workerParams) {
    override suspend fun doWork(): Result {
        Log.i(TokenWorker::class.java.name, "Start background refresh token")
        AnalysisUtil.log("Start background refresh token")
        val service = NaviToTeslaService(applicationContext)
        val token = service.refreshToken()
        return if (token != null) {
            Result.success()
        } else {
            Result.failure()
        }
    }

    companion object {
        private const val WORKER_NAME = "refreshTokenWorker"

        fun startBackgroundWork(context: Context) {
            CoroutineScope(Dispatchers.Main).launch {
                if (PreferencesUtil.loadToken() == null) {
                    AnalysisUtil.log("Token is empty. add token refresh work ignore")
                    return@launch
                }
                AnalysisUtil.log("Add background refresh token")
                val workRequest: PeriodicWorkRequest =
                    PeriodicWorkRequestBuilder<TokenWorker>(350, TimeUnit.MINUTES)
                        .setConstraints(
                            Constraints
                                .Builder()
                                .setRequiredNetworkType(NetworkType.CONNECTED)
                                .build(),
                        ).build()
                WorkManager.getInstance(context).enqueueUniquePeriodicWork(
                    WORKER_NAME,
                    ExistingPeriodicWorkPolicy.KEEP,
                    workRequest,
                )
            }
        }

        fun cancelBackgroundWork(context: Context) {
            try {
                WorkManager.getInstance(context).cancelUniqueWork(WORKER_NAME)
            } catch (e: Exception) {
                AnalysisUtil.recordException(e)
            }
        }
    }
}
